### PR TITLE
Module 2

### PR DESCRIPTION
**Pull request by JoLeeSoftwire**
Including descriptions on tasks

### CODE REVIEW

File: session_items.py [hunk 1]
```diff
@@ -1,73 +1,111 @@
 from flask import session
```
Comment (Jonesey13) [inline session_items.py]: We don't use this anymore (which means the name of this file is obsolete). Probably worth renaming to `trello_api.py` or something similar.

File: session_items.py [hunk 2]
```diff
@@ -1,73 +1,111 @@
...
 _DEFAULT_ITEMS = [
     { 'id': 1, 'status': 'Not Started', 'title': 'List saved todo items' },
     { 'id': 2, 'status': 'Not Started', 'title': 'Allow new items to be added' }
 ]
 
+BOARD_ID = os.environ.get('BOARD_ID')
+TRELLO_KEY = os.environ.get("TRELLO_KEY")
+TRELLO_TOKEN = os.environ.get("TRELLO_TOKEN")
+
+TODO_LIST_ID = os.environ.get("TODO_LIST_ID")
```
Comment (Jonesey13) [inline session_items.py]: These two could be inferred from the list names

File: session_items.py [hunk 3]
```diff
@@ -1,73 +1,111 @@
...
+import requests
+
+from Task import Task, Status
 
 _DEFAULT_ITEMS = [
     { 'id': 1, 'status': 'Not Started', 'title': 'List saved todo items' },
     { 'id': 2, 'status': 'Not Started', 'title': 'Allow new items to be added' }
 ]
 
+BOARD_ID = os.environ.get('BOARD_ID')
```
Comment (Jonesey13) [inline session_items.py]: Is this ever used?

File: session_items.py [hunk 4]
```diff
@@ -1,73 +1,111 @@
...
+TRELLO_KEY = os.environ.get("TRELLO_KEY")
+TRELLO_TOKEN = os.environ.get("TRELLO_TOKEN")
+
+TODO_LIST_ID = os.environ.get("TODO_LIST_ID")
+DONE_LIST_ID = os.environ.get("DONE_LIST_ID")
+
+trello_baseurl = "https://api.trello.com/"
+trello_apiversion = '1'
+trello_commonurl = trello_baseurl + trello_apiversion
+boardselector = "/board/"
```
Comment (Jonesey13) [inline session_items.py]: Is this ever used?

File: session_items.py [hunk 5]
```diff
@@ -1,73 +1,111 @@
...
+statusToID = {
+    Status.ToDo: TODO_LIST_ID,
+    Status.Done: DONE_LIST_ID
+}
+IDToStatus = {
+    TODO_LIST_ID: Status.ToDo,
+    DONE_LIST_ID: Status.Done
+}
 
 def get_items():
```
Comment (Jonesey13) [inline session_items.py]: These methods could be `@staticmethod`s on a TrelloApi class.

File: session_items.py [hunk 6]
```diff
@@ -1,73 +1,111 @@
...
+TRELLO_TOKEN = os.environ.get("TRELLO_TOKEN")
+
+TODO_LIST_ID = os.environ.get("TODO_LIST_ID")
+DONE_LIST_ID = os.environ.get("DONE_LIST_ID")
+
+trello_baseurl = "https://api.trello.com/"
+trello_apiversion = '1'
+trello_commonurl = trello_baseurl + trello_apiversion
+boardselector = "/board/"
+listselector = "/lists/"
```
Comment (Jonesey13) [inline session_items.py]: Given these 2 are only ever used once I'd be tempted just to use string interpolation to keep things readable (i.e. `f'...'`)

File: session_items.py [hunk 7]
```diff
@@ -1,73 +1,111 @@
...
+    endpoint = trello_commonurl + cardselector + task_id
+    extraparams = {
+        "idList": DONE_LIST_ID
+    }
+    allparams = custom_query_params(extraparams)
+
+    try:
+        trelloDone = requests.put(endpoint, params=allparams).json()
+    except:
+        print("card with id "+str(id)+" not found, will refresh")
```
Comment (Jonesey13) [inline session_items.py]: String interpolation is your friend here.

File: session_items.py [hunk 8]
```diff
@@ -1,73 +1,111 @@
...
+    
+    todo = get_items_with_satus(Status.ToDo)
+    done = get_items_with_satus(Status.Done)
 
-def get_sorted_items():
-    items = get_items()
-    return sorted(items, key=lambda item: item["status"])
+    return todo + done
 
+def get_items_with_satus(status):
```
Comment (Jonesey13) [inline session_items.py]: typo in `satus`

File: session_items.py [hunk 9]
```diff
@@ -1,73 +1,111 @@
 from flask import session
+import os
+import requests
+
+from Task import Task, Status
 
 _DEFAULT_ITEMS = [
```
Comment (Jonesey13) [inline session_items.py]: Unused

File: session_items.py [hunk 10]
```diff
@@ -1,73 +1,111 @@
...
+def get_items_with_satus(status):
+    endpoint = trello_commonurl + listselector + statusToID[status] + cardselector
+    trelloTasks = requests.get(endpoint, params=default_query_params).json()
 
-def get_item(id):
-    """
-    Fetches the saved item with the specified ID.
+    tasks = list()
+    for item in trelloTasks:
+        taskObj = Task(item['id'], item['name'], status, item['desc'])
```
Comment (Jonesey13) [inline session_items.py]: Could be moved into a `@classmethod` conversion method on the Task class (e.g. `from_card(item, status)`).

File: app.py [hunk 11]
```diff
@@ -6,20 +6,19 @@
```
Comment (Jonesey13) [inline app.py:6]: Lines 2 & 5 are obsolete (also the SECRET_KEY can be removed)

**Comment by Jonesey13** [comment]
`requirements.txt` is missing a dependency (`requests`)

**Comment by Jonesey13** [comment]
`.env.template` is missing some values

**Review by Jonesey13 — CHANGES_REQUESTED** [review]
Very nearly there! (just missing a dependency to get the project running).

Apart from that I've just left some refactoring comments.

File: app.py (hunk 11 above)
Comment (JoLeeSoftwire) [reply app.py:6]: What did SECRET_KEY used to do? was that for identifying/accessing the session cookies/data?
also I'm not sure line 2 is obsolete, did you mean a different line?

File: session_items.py (hunk 6 above)
Comment (JoLeeSoftwire) [reply session_items.py]: I think I prefer the consistency of doing it the same for `/cards/` `/lists/` and `/board/` , and `/cards/` comes up a few times

File: session_items.py (hunk 5 above)
Comment (JoLeeSoftwire) [reply session_items.py]: When i try to set that up I get `AttributeError: module 'TrelloApi' has no attribute 'get_items'` at runtime, and I can't see why, I haven't renamed the function, I've marked it as static
I'm not quite sure the benefit of it either, if every method is static. Is it purely to be consistently object-oriented?

File: session_items.py (hunk 2 above)
Comment (JoLeeSoftwire) [reply session_items.py]: Can a have a method which runs whenever the app is launched, which retrieves and stores them only if it doesn't already have them? sort of like an init of the whole app? It seems silly to retrieve them separately every time, and seems like it'd clutter the code check for them on every method and fetch if we don't have (even knowing that'd be extracted out to a shared function)

File: app.py (hunk 11 above)
Comment (Jonesey13) [reply app.py:6]: SECRET_KEY was used to encrypt the flask session cookie, but we're using it no more: https://flask.palletsprojects.com/en/1.1.x/config/#SECRET_KEY

You're right, we still need line 2 (just renamed). Line 5 can go though (along with `flask_config.py`)

File: session_items.py (hunk 5 above)
Comment (Jonesey13) [reply session_items.py]: Essentially you're right, it's really only for namespacing the method under a class (which some might argue is better than using the file as a namespace).

Comment (Jonesey13) [reply session_items.py]: You'd have to send me a code snippet for me to help you with that error though.

File: session_items.py (hunk 2 above)
Comment (Jonesey13) [reply session_items.py]: I'd advocate keeping the webserver as stateless as possible at this point. Later on we can introduce database and caching layers to work around this, but I'd argue that's overkill for this exercise.

**Review by Jonesey13 — APPROVED** [review]
Dependency issue fixed so approving this for merge. 👍

**Pull request merged**